### PR TITLE
Add llama3 support on llama.cpp for non-apple silicon devices and benchmark results

### PR DESCRIPTION
**Pull request by Char15Xu**

### CODE REVIEW

**Comment by zenodflow** [comment]
@Char15Xu This is nice. Could you first rebase your branch to the current main? 

I did a bunch of updates over feat-local in https://github.com/abcsys/libem/pull/40.

**Comment by Char15Xu** [comment]
it should be good now.

File: Makefile [hunk 1]
```diff
@@ -19,7 +19,7 @@ chat:
 	python examples/tool/chat.py
 block:
 	python examples/block.py
-all: match browse chat
+all: match match-local browse chat
```
Comment (zenodflow) [inline Makefile:22]: I think I've removed this in #40

Comment (zenodflow) [reply Makefile:22]: Also, pls remove .DS_Store from the commit. It should be in the `.gitignore` now.

File: Makefile [hunk 2]
```diff
@@ -50,6 +50,8 @@ batch: async_batch
 .PHONY: mlx_lm llama local
 mlx_lm:
 	pip install mlx_lm
+llama.cpp:
```
Comment (zenodflow) [inline Makefile:53]: llama_cpp

File: examples/local.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import libem
```
Comment (zenodflow) [inline examples/local.py:1]: See if this is a duplicate of https://github.com/abcsys/libem/blob/main/examples/model/llama.py

File: libem/core/model/llama.py [hunk 4]
```diff
@@ -31,20 +31,22 @@ def call(prompt: str | list | dict,
...
-    EOS = "<|eot_id|>"
-
-    input_text = BOS
-    for message in messages:
-        if message['role'] == 'system':
-            input_text = input_text + SYS + message['content'] + EOS
-        elif message['role'] == 'user':
-            input_text = input_text + USER + message['content'] + EOS
-    input_text = input_text + ASSIS
+    message_openai = (messages or []) + (context or [])
```
Comment (zenodflow) [inline libem/core/model/llama.py:34]: why `message_openai`? This message is in the llama.py module

File: libem/core/model/llama.py [hunk 5]
```diff
@@ -74,12 +76,43 @@ def call(prompt: str | list | dict,
             raise libem.ToolUseUnsupported("Tool use is not supported")
 
         response = generate(model, tokenizer, prompt=messages[0], temp=temperature)
+        response_message = "messages is not supported"
+        
     else:
-        raise ValueError(f"{platform.machine()} {platform.system()} is not supported.")
+        if model == "llama3":
```
Comment (zenodflow) [inline libem/core/model/llama.py:82]: reverse the model serial number and the platform `if .. else` clause -- i.e. put the `if model == "llama"` the top-most if clause

File: libem/core/model/openai.py [hunk 6]
```diff
@@ -92,7 +92,7 @@ async def async_call(
             raise libem.ModelTimedoutException(e)
 
         response_message = response.choices[0].message
-
+        print(response_message)
```
Comment (zenodflow) [inline libem/core/model/openai.py:95]: kill it

**Review by zenodflow — CHANGES_REQUESTED** [review]
Solid work @Char15Xu! Just some minor comments and we are good to go!

File: libem/core/model/llama.py (hunk 4 above)
Comment (Char15Xu) [reply libem/core/model/llama.py:34]: chatCompletion in llama.cpp takes in a message of the openai format

**Review by zenodflow — APPROVED** [review]

**Pull request merged**